File: 04-vector-search/python/SearchComponents/searches.py
import SearchComponents.vectorSearch as VectorSearch
import SearchComponents.completion as Completion
import os


def Run_vector_search(embeddings_deployment, AzureOpenAIClient, client, cosmos_db_mongodb_database):
    os.system('cls' if os.name == 'nt' else 'clear')
    print("What would you like to know about our bike shop's inventory?")
    user_input = input("Prompt: ")
    maxResults = 20
    vector_column = "productVector" # input("Enter the vector column name: ")
    collection_name = "products" # input("Enter the collection name: ")


    db = client[cosmos_db_mongodb_database]
    collection = db[collection_name]
    results = VectorSearch.vector_search(user_input, vector_column, collection, embeddings_deployment, AzureOpenAIClient, maxResults)
    for result in results: 
        print(f"Similarity Score: {result['similarityScore']}"
              + f", category: {result['document']['categoryName']}" 
              + f", Product: {result['document']['name']}")  


def Run_GPT_search(embeddings_deployment, AzureOpenAIClient, completion_deployment, AzureOpenAICompletionClient, client, cosmos_db_mongodb_database):
    maxResults = 20
    vector_column = "productVector" # input("Enter the vector column name: ")
    collection_name = "products" # input("Enter the collection name: ")

    db = client[cosmos_db_mongodb_database]
    collection = db[collection_name]

    user_input = ""
    os.system('cls' if os.name == 'nt' else 'clear')
    print("What would you like to ask about our bike shop's inventory? Type 'end' to end the session. ")
    user_input = input("Prompt: ")
    while user_input.lower() != "end":
        results_for_prompt = VectorSearch.vector_search(user_input, vector_column, collection, embeddings_deployment, AzureOpenAIClient, maxResults)

        completions_results = Completion.generate_completion(results_for_prompt,completion_deployment,AzureOpenAICompletionClient,user_input)
        print("\n"+completions_results['choices'][0]['message']['content'])

        print("\nWhat would you like to ask about our bike shop's inventory? Type 'end' to end the session. ")
        user_input = input("Prompt: ")

File: 04-vector-search/Solution/node.js/SearchComponents/searches.js
// Import the required modules
const VectorSearch = require('../SearchComponents/vectorSearch');
const Completion = require('../SearchComponents/completion');

// Asynchronous function to run a vector search
async function runVectorSearch(embeddingsDeployment, AzureOpenAIClient, client, cosmosDbMongodbDatabase, rl) {
    // Clear the console and ask the user for input
    console.clear();
    console.log("What would you like to know about our bike shop's inventory?");
    // read the user input for a new prompt
    const userInput = await new Promise(resolve => rl.question("Prompt: ", resolve));
    // Define the maximum number of results, the vector column, and the collection name
    const maxResults = 20;
    const vectorColumn = "productVector";
    const collectionName = "products";

    // Connect to the database and get the collection
    const db = client.db(cosmosDbMongodbDatabase);
    const collection = db.collection(collectionName);
    // Run the vector search and print the results
    const results = await VectorSearch.vectorSearch(userInput, vectorColumn, collection, embeddingsDeployment, AzureOpenAIClient, maxResults);
    for (let result of results) {
        console.log(`Similarity Score: ${result.similarityScore}, category: ${result.document.categoryName}, Product: ${result.document.name}`);
    }
}

// Asynchronous function to run a GPT-3 search
async function runGPTSearch(embeddingsDeployment, AzureOpenAIClient, completionDeployment, client, cosmosDbMongodbDatabase, rl) {
    // Define the maximum number of results, the vector column, and the collection name
    const maxResults = 20;
    const vectorColumn = "productVector";
    const collectionName = "products";

    // Connect to the database and get the collection
    const db = client.db(cosmosDbMongodbDatabase);
    const collection = db.collection(collectionName);

    // Initialize the user input variable
    let userInput = "";
    // Clear the console and ask the user for input
    console.clear();
    console.log("What would you like to ask about our bike shop's inventory? Type 'end' to end the session. ");
    userInput = await new Promise(resolve => rl.question("Prompt: ", resolve));
    // Continue asking for input until the user types 'end'
    while (userInput.toLowerCase() !== "end") {
        // Run the vector search
        const resultsForPrompt = await VectorSearch.vectorSearch(userInput, vectorColumn, collection, embeddingsDeployment, AzureOpenAIClient, maxResults);

        // Generate completions based on the vector search results
        const completionsResults = await Completion.generateCompletion(resultsForPrompt, completionDeployment, AzureOpenAIClient, userInput);
        // Print the first completion result
        console.log("\n" + completionsResults.choices[0].message.content);

        // Ask the user for more input
        console.log("\nWhat would you like to ask about our bike shop's inventory? Type 'end' to end the session. ");
        userInput = await new Promise(resolve => rl.question("Prompt: ", resolve));
    }
}

// Export the runVectorSearch and runGPTSearch functions
module.exports.runVectorSearch = runVectorSearch;
module.exports.runGPTSearch = runGPTSearch;